csi/controller.go

```diff
@@ -455,9 +471,122 @@ func csiRequestsSharedVolume(req *csi.CreateVolumeRequest) bool {
 	return false
 }
 
-/*
-For next patches what still needs to be worked on in the Conroller server:
+// CreateSnapshot is a CSI implementation to create a snapshot from the volume
+func (o *OsdCsiServer) CreateSnapshot(
+	ctx context.Context,
+	req *csi.CreateSnapshotRequest,
+) (*csi.CreateSnapshotResponse, error) {
 
-	GetCapacity(context.Context, *GetCapacityRequest) (*GetCapacityResponse, error)
-	ControllerGetCapabilities(context.Context, *ControllerGetCapabilitiesRequest) (*ControllerGetCapabilitiesResponse, error)
-*/
+	if len(req.GetSourceVolumeId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume id must be provided")
+	} else if len(req.GetName()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Name must be provided")
+	}
+
+	// Check if the snapshot with this name already exists
+	v, err := util.VolumeFromName(o.driver, req.GetName())
+	if err == nil {
+		// Verify the parent is the same
+		if req.GetSourceVolumeId() != v.GetSource().GetParent() {
+			return nil, status.Error(codes.AlreadyExists, "Requested snapshot already exists for another source volume id")
+		}
+
+		// Return current snapshot info
+		createdAt, err := ptypes.Timestamp(v.GetCtime())
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Failed to get time snapshot was created: %v", err)
+		}
+		return &csi.CreateSnapshotResponse{
+			Snapshot: &csi.Snapshot{
+				Id:             v.GetId(),
+				SourceVolumeId: v.GetSource().GetParent(),
+				CreatedAt:      createdAt.Unix(),
+				Status: &csi.SnapshotStatus{
+					// This means that we are not uploading our snapshot
+					// We may add support flow cloud snaps in future patches
+					Type: csi.SnapshotStatus_READY,
+				},
+			},
+		}, nil
+	}
```

Comment:
Shouldn't we check for err != nil and see if the error is NotFound?

Reply:
I think we should fix https://github.com/libopenstorage/openstorage/blob/master/pkg/util/volume.go#L38 so that it returns an exported named error value. Then we should add it here and in other areas. 

I created #631 to track this.